File: sdks/ruby/test/testing/mock_context_test.rb
# frozen_string_literal: true

require_relative "../test_helper"

class MockContextTest < Minitest::Test
  def test_default_values
    ctx = Nanoservice::Testing::MockContext.new.build

    assert_equal "test-execution-id", ctx.id
    assert_equal "test-workflow", ctx.workflow_name
    assert_equal "/workflows/test", ctx.workflow_path
    assert_equal "POST", ctx.request.method
    assert_equal "/test", ctx.request.url
    assert_equal({}, ctx.vars)
    assert_equal({}, ctx.env)
  end

  def test_with_id
    ctx = Nanoservice::Testing::MockContext.new
      .with_id("custom-id")
      .build

    assert_equal "custom-id", ctx.id
  end

  def test_with_workflow
    ctx = Nanoservice::Testing::MockContext.new
      .with_workflow("my-workflow", "/wf/my-workflow")
      .build

    assert_equal "my-workflow", ctx.workflow_name
    assert_equal "/wf/my-workflow", ctx.workflow_path
  end

  def test_with_workflow_default_path
    ctx = Nanoservice::Testing::MockContext.new
      .with_workflow("auth")
      .build

    assert_equal "auth", ctx.workflow_name
    assert_equal "/workflows/auth", ctx.workflow_path
  end

  def test_with_body
    ctx = Nanoservice::Testing::MockContext.new
      .with_body({ "name" => "John", "age" => 30 })
      .build

    assert_equal "John", ctx.request.body["name"]
    assert_equal 30, ctx.request.body["age"]
  end

  def test_with_headers
    ctx = Nanoservice::Testing::MockContext.new
      .with_headers({ "Authorization" => "Bearer token" })
      .build

    assert_equal "Bearer token", ctx.request.headers["Authorization"]
  end

  def test_with_method
    ctx = Nanoservice::Testing::MockContext.new
      .with_method("GET")
      .build

    assert_equal "GET", ctx.request.method
  end

  def test_with_var
    ctx = Nanoservice::Testing::MockContext.new
      .with_var("key1", "value1")
      .with_var("key2", 42)
      .build

    assert_equal "value1", ctx.vars["key1"]
    assert_equal 42, ctx.vars["key2"]
  end

  def test_with_env
    ctx = Nanoservice::Testing::MockContext.new
      .with_env("API_KEY", "secret123")
      .with_env("DEBUG", "true")
      .build

    assert_equal "secret123", ctx.env["API_KEY"]
    assert_equal "true", ctx.env["DEBUG"]
  end

  def test_fluent_chaining
    ctx = Nanoservice::Testing::MockContext.new
      .with_id("chain-test")
      .with_workflow("wf")
      .with_body({ "x" => 1 })
      .with_headers({ "H" => "V" })
      .with_var("a", "b")
      .with_env("E", "F")
      .build

    assert_equal "chain-test", ctx.id
    assert_equal "wf", ctx.workflow_name
    assert_equal 1, ctx.request.body["x"]
    assert_equal "V", ctx.request.headers["H"]
    assert_equal "b", ctx.vars["a"]
    assert_equal "F", ctx.env["E"]
  end

  def test_context_set_and_get_var
    ctx = Nanoservice::Testing::MockContext.new.build

    ctx.set_var("greeting", "Hello!")
    assert_equal "Hello!", ctx.get_var("greeting")
    assert_equal "Hello!", ctx.get_var_str("greeting")
    assert_nil ctx.get_var("missing")
    assert_nil ctx.get_var_str("missing")
  end

  def test_context_get_var_str_non_string
    ctx = Nanoservice::Testing::MockContext.new
      .with_var("count", 42)
      .build

    assert_equal 42, ctx.get_var("count")
    assert_nil ctx.get_var_str("count")
  end

  def test_context_to_hash_roundtrip
    ctx = Nanoservice::Testing::MockContext.new
      .with_id("roundtrip")
      .with_body({ "key" => "val" })
      .with_var("v", "1")
      .build

    hash     = ctx.to_hash
    restored = Nanoservice::Types::Context.from_hash(hash)

    assert_equal "roundtrip", restored.id
    assert_equal "val", restored.request.body["key"]
    assert_equal "1", restored.vars["v"]
  end

  def test_request_body_str
    ctx = Nanoservice::Testing::MockContext.new
      .with_body({ "name" => "Alice", "count" => 5 })
      .build

    assert_equal "Alice", ctx.request.body_str("name")
    assert_nil ctx.request.body_str("count")  # Not a string
    assert_nil ctx.request.body_str("missing")
  end
end

File: sdks/php/tests/Testing/MockContextTest.php
<?php

declare(strict_types=1);

namespace Blok\Nanoservice\Tests\Testing;

use Blok\Nanoservice\Testing\MockContext;
use PHPUnit\Framework\TestCase;

final class MockContextTest extends TestCase
{
    public function testDefaultValues(): void
    {
        $ctx = MockContext::create()->build();

        $this->assertSame('test-execution-id', $ctx->id);
        $this->assertSame('test-workflow', $ctx->workflowName);
        $this->assertSame('/workflows/test', $ctx->workflowPath);
        $this->assertSame('POST', $ctx->request->method);
        $this->assertSame('/test', $ctx->request->url);
        $this->assertSame('http://localhost:8080', $ctx->request->baseUrl);
        $this->assertEmpty($ctx->vars);
        $this->assertEmpty($ctx->env);
    }

    public function testWithId(): void
    {
        $ctx = MockContext::create()
            ->withId('custom-id')
            ->build();

        $this->assertSame('custom-id', $ctx->id);
    }

    public function testWithWorkflow(): void
    {
        $ctx = MockContext::create()
            ->withWorkflow('my-workflow', '/workflows/my')
            ->build();

        $this->assertSame('my-workflow', $ctx->workflowName);
        $this->assertSame('/workflows/my', $ctx->workflowPath);
    }

    public function testWithBody(): void
    {
        $ctx = MockContext::create()
            ->withBody(['name' => 'test', 'value' => 42])
            ->build();

        $this->assertSame(['name' => 'test', 'value' => 42], $ctx->request->body);
        $this->assertSame('test', $ctx->request->bodyStr('name'));
    }

    public function testWithHeaders(): void
    {
        $ctx = MockContext::create()
            ->withHeaders(['Authorization' => 'Bearer token', 'Accept' => 'application/json'])
            ->build();

        $this->assertSame('Bearer token', $ctx->request->headers['Authorization']);
        $this->assertSame('application/json', $ctx->request->headers['Accept']);
    }

    public function testWithMethod(): void
    {
        $ctx = MockContext::create()
            ->withMethod('GET')
            ->build();

        $this->assertSame('GET', $ctx->request->method);
    }

    public function testWithUrl(): void
    {
        $ctx = MockContext::create()
            ->withUrl('/api/users')
            ->build();

        $this->assertSame('/api/users', $ctx->request->url);
    }

    public function testWithVar(): void
    {
        $ctx = MockContext::create()
            ->withVar('key1', 'value1')
            ->withVar('key2', ['nested' => true])
            ->build();

        $this->assertSame('value1', $ctx->getVar('key1'));
        $this->assertSame(['nested' => true], $ctx->getVar('key2'));
        $this->assertSame('value1', $ctx->getVarStr('key1'));
    }

    public function testWithEnv(): void
    {
        $ctx = MockContext::create()
            ->withEnv('API_KEY', 'secret-key')
            ->withEnv('DEBUG', 'true')
            ->build();

        $this->assertSame('secret-key', $ctx->env['API_KEY']);
        $this->assertSame('true', $ctx->env['DEBUG']);
    }

    public function testWithParams(): void
    {
        $ctx = MockContext::create()
            ->withParams(['id' => '123'])
            ->build();

        $this->assertSame('123', $ctx->request->params['id']);
    }

    public function testWithQuery(): void
    {
        $ctx = MockContext::create()
            ->withQuery(['page' => '1', 'limit' => '10'])
            ->build();

        $this->assertSame('1', $ctx->request->query['page']);
        $this->assertSame('10', $ctx->request->query['limit']);
    }

    public function testFluentChaining(): void
    {
        $ctx = MockContext::create()
            ->withId('chain-test')
            ->withWorkflow('wf', '/wf')
            ->withBody(['name' => 'test'])
            ->withHeaders(['X-Custom' => 'yes'])
            ->withMethod('PUT')
            ->withVar('prev', 'result')
            ->withEnv('MODE', 'test')
            ->build();

        $this->assertSame('chain-test', $ctx->id);
        $this->assertSame('wf', $ctx->workflowName);
        $this->assertSame(['name' => 'test'], $ctx->request->body);
        $this->assertSame('yes', $ctx->request->headers['X-Custom']);
        $this->assertSame('PUT', $ctx->request->method);
        $this->assertSame('result', $ctx->getVarStr('prev'));
        $this->assertSame('test', $ctx->env['MODE']);
    }

    public function testContextSetAndGetVar(): void
    {
        $ctx = MockContext::create()->build();

        $ctx->setVar('key', 'value');
        $this->assertSame('value', $ctx->getVar('key'));
        $this->assertSame('value', $ctx->getVarStr('key'));
        $this->assertNull($ctx->getVar('missing'));
        $this->assertNull($ctx->getVarStr('missing'));
    }

    public function testContextGetVarStrReturnsNullForNonString(): void
    {
        $ctx = MockContext::create()
            ->withVar('number', 42)
            ->build();

        $this->assertSame(42, $ctx->getVar('number'));
        $this->assertNull($ctx->getVarStr('number'));
    }

    public function testContextToArrayAndFromArray(): void
    {
        $ctx = MockContext::create()
            ->withId('round-trip')
            ->withWorkflow('test-wf', '/test')
            ->withBody(['name' => 'test'])
            ->withVar('key', 'val')
            ->build();

        $array = $ctx->toArray();

        $this->assertSame('round-trip', $array['id']);
        $this->assertSame('test-wf', $array['workflow_name']);
        $this->assertSame('/test', $array['workflow_path']);
        $this->assertSame(['name' => 'test'], $array['request']['body']);
        $this->assertSame('val', $array['vars']['key']);
    }
}
